File: ios/Classes/Models/Mpgs/MpgsCard.swift
//
//  MpgsCard.swift
//  trywithmpgs
//
//  Created by Ray Chow on 21/4/2022.
//

import Foundation

struct MpgsCard: Mappable {
    let name: String
    let number: String
    let cvc: String
    let expirationMonth: Int
    let expirationYear: Int
    
    static func mapFromDict(dict: [String : Any]) -> MpgsCard? {
        guard let name = dict[MethodChannelCardInformationKey.nameOnCard.rawValue] as? String,
              let number = dict[MethodChannelCardInformationKey.cardNumber.rawValue] as? String,
              let cvc = dict[MethodChannelCardInformationKey.cvc.rawValue] as? String,
              let expirationMonth = dict[MethodChannelCardInformationKey.expirationMonth.rawValue] as? Int,
              let expirationYear = dict[MethodChannelCardInformationKey.expirationYear.rawValue] as? Int
        else { return nil }
        
        return MpgsCard(name: name,
                        number: number,
                        cvc: cvc,
                        expirationMonth: expirationMonth,
                        expirationYear: expirationYear)
    }
}

File: android/src/main/kotlin/com/keychainpay/keychainpay_mpgs/models/mpgs/MpgsCard.kt
package com.keychainpay.keychainpay_mpgs.models.mpgs

import com.keychainpay.keychainpay_mpgs.enums.method_channel.card_merchant.common.card_info.MethodChannelCardInformationKey
import com.keychainpay.keychainpay_mpgs.models.base.Mappable
import com.keychainpay.keychainpay_mpgs.utils.safeLet

class MpgsCard(val name: String,
               val number: String,
               val cvc: String,
               val expirationMonth: Int,
               val expirationYear: Int) {

    companion object: Mappable<MpgsCard> {
        override fun mapFromHashMap(hashMap: HashMap<String, Any>): MpgsCard? {
            return safeLet(hashMap[MethodChannelCardInformationKey.NAME_ON_CARD.key] as? String,
                hashMap[MethodChannelCardInformationKey.CARD_NUMBER.key] as? String,
                hashMap[MethodChannelCardInformationKey.CVC.key] as? String,
                hashMap[MethodChannelCardInformationKey.EXPIRATION_MONTH.key] as? Int,
                hashMap[MethodChannelCardInformationKey.EXPIRATION_YEAR.key] as? Int) {
                name, number, cvc, expirationMonth, expirationYear ->
                MpgsCard(name, number, cvc, expirationMonth, expirationYear)
            }
        }
    }
}